### PR TITLE
Make `check` comply with Hackage requirements

### PR DESCRIPTION
**Pull request by ffaf1**
tl;dr: `cabal check` will exit with `1` if and only if `Hackage would reject the package`.

As today, `cabal check` handles reports a “failure” in two distinct ways:
- first one: with an informative `Hackage would reject this package.` string
- second one: with exitcode `1`. This one is *not* the same as `Hackage would reject this package.` It is a bit more strict: you could end up with `exitcode 1` but no `Hackage would reject this package.` string.

What this patch does: make the two “failures” bijective. You will get `exitcode 1` iif `Hackage would reject this package.`

Examples of checks which will *not* exit with `1` anymore:
- `description` longer than `synopsis`;
- `-O1` usage;
- missing `maintainer`.

Check last commit for all of them. If someone has a good reason for them to return `1`, let me know (would CI get more annoying to set up? Third-party tools? Maybe it is better to return multiple failure exitcodes?).  

Towards: #8880
Reviewers: review by commit to skip testsuite changes. 
Related from an UX point of view: #8587  

---

QA notes:
- run `cabal check` in any project of yours
- take note if you see `Hackage would reject this package` in the output.
- take note of the exit code (`printf '%d\n' $?`).
- exit code should be `1` if and only if you see the message `Hackage would reject this package.`

(a quick way to get `Hackage would refuse this package` string is to remove `default-language:   Haskell2010` from your `.cabal`)

Specific to this patch:
- modify your package `synopsis` to be longer than your `description`
- exit code should be `0` and you should not see any `Hackage would reject this package.` message
 
---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.

Tests: I have adjusted the testsuite appropriately.

### CODE REVIEW

**Review by gbaz — APPROVED** [review]

File: Cabal/src/Distribution/PackageDescription/Check.hs [hunk 1]
```diff
@@ -846,6 +847,12 @@ data PackageCheck =
      | PackageDistInexcusable { explanation :: CheckExplanation }
   deriving (Eq, Ord)
 
+-- | Would Hackage refuse a package because of this error?
+isHackageDistError :: PackageCheck -> Bool
+isHackageDistError (PackageDistSuspicious {}) = False
+isHackageDistError (PackageDistSuspiciousWarn {}) = False
+isHackageDistError _ = True
```
Comment (andreasabel) [inline Cabal/src/Distribution/PackageDescription/Check.hs:854]: (Without having looked at the `PackageCheck` constructors:)
Can we get the catch-all expanded into the constructors?  (Forces future extensions of `PackageCheck` to revisit this predicate.)

**Review by andreasabel — COMMENTED** [review]
Great feature!

File: Cabal/src/Distribution/PackageDescription/Check.hs (hunk 1 above)
Comment (ffaf1) [reply Cabal/src/Distribution/PackageDescription/Check.hs]: Sensible reasoning, added!

Comment (ffaf1) [reply Cabal/src/Distribution/PackageDescription/Check.hs]: @andreasabel was that your only suggestion? If so, if you could approve it, I have another PR ready to close #8880.

File: doc/cabal-commands.rst [hunk 2]
```diff
@@ -977,7 +977,8 @@ Run ``cabal check`` in the folder where your ``.cabal`` package file is.
     Set verbosity level (0–3, default is 1).
 
 ``cabal check`` mimics Hackage's requirements: if no error or warning
-is reported, Hackage should accept your package.
+is reported, Hackage should accept your package. If ``cabal check`` exits
+with ``1``, Hackage will refuse it.
```
Comment (ulysses4ever) [inline doc/cabal-commands.rst:981]: There's still a grey area but stressed in the text: could it be that errors (i assume, not) or warnings (maybe?) are reported but check record with 0? If it can what does it mean? Anyway, i think it'd be best to address this case explicitly in the manual.

File: Cabal/src/Distribution/PackageDescription/Check.hs [hunk 3]
```diff
@@ -846,6 +847,14 @@ data PackageCheck =
      | PackageDistInexcusable { explanation :: CheckExplanation }
   deriving (Eq, Ord)
 
+-- | Would Hackage refuse a package because of this error?
+isHackageDistError :: PackageCheck -> Bool
+isHackageDistError (PackageBuildImpossible {}) = True
```
Comment (ulysses4ever) [inline Cabal/src/Distribution/PackageDescription/Check.hs:852]: I usually use lambda-case in such cases, so that you don't have to repeat the function name numerous times. But up to you...

**Review by ulysses4ever — APPROVED** [review]
Tremendous! Thanks!

File: doc/cabal-commands.rst (hunk 2 above)
Comment (ffaf1) [reply doc/cabal-commands.rst]: Indeed it was not clear. New text:

---

``cabal check`` mimics Hackage's requirements: if no error or warning
is reported, Hackage should accept your package. If errors are present      
``cabal check`` exits with ``1`` and Hackage will will refuse the package.

---

File: doc/cabal-commands.rst [hunk 4]
```diff
@@ -977,7 +977,8 @@ Run ``cabal check`` in the folder where your ``.cabal`` package file is.
     Set verbosity level (0–3, default is 1).
 
 ``cabal check`` mimics Hackage's requirements: if no error or warning
-is reported, Hackage should accept your package.
+is reported, Hackage should accept your package. If errors are present
+``cabal check`` exits with ``1`` and Hackage will will refuse the package.
```
Comment (ulysses4ever) [inline doc/cabal-commands.rst:981]:
```suggestion
``cabal check`` exits with ``1`` and Hackage will refuse the package.
```

File: doc/cabal-commands.rst (hunk 2 above)
Comment (ulysses4ever) [reply doc/cabal-commands.rst]: It does, thank you!

Sorry, there was a typo and i meant to create a suggestion but accidentally committed it...
